### ISSUE TITLE
Reducing number of NTTs in some cases

### ISSUE THREAD

**Issue by mschof** [Plonky3/Plonky3#535]
Depending on the setting, various relations in a trace table can be purely linear between different columns (e.g., this may happen for limb decompositions). In this case, it's possible to omit computing the NTT of one of the columns because it's just a linear combination of the previous NTTs. This is somewhat similar to the idea behind virtual commitments/polynomials.

In particular, consider a Monolith-31 trace where 31-bit values are decomposed into two 16-bit limbs. Then, for every limb decomposition, 1/3 of the NTTs can be omitted.

Does this optimization make sense?

**Comment by dlubarov** [comment]
Hey thanks for the suggestion! It makes sense, but it just gets a bit more complex/naunced when considering that we also reuse the same RS codewords as FRI inputs. So if we stick with a non-systematic RS code, we still need 3 NTTs to compute FRI inputs.

If ZK is disabled, we could consider using FRI with a systematic RS code instead. I think mixing systematic and non-systematic inputs could be a possibility too - since low-degreeness shouldn't be affected by coset shifts, FRI could pretend all inputs are systematic.

Would Monolith work well with an AIR that's entirely (max) degree 2? It would probably be easier for us if we can handle all polynomials within an AIR the same way, vs identifying certain polynomials (not involved in degree 3 constraints) to handle differently.

**Comment by mschof** [comment]
Probably I'm missing something, but the polynomial addition (coefficient addition) would lead to the final interpolating polynomial anyway, which should be equivalent to computing the NTT.

Anyway, yes: Monolith is designed to work with degree-2 constraints by default. The degree of the round function is actually 2, independent of the prime number being used (that's different than in Poseidon etc.).

**Comment by dlubarov** [comment]
Hm that's true, I was thinking of a column with degree 2 rather than degree 1 constraints. If a column is really degree 1 only, i.e. it can be written as a linear combination of others, shouldn't it be possible to remove the column?

I think there may be some "convenience" type reasons to keep such a column, like some lookup implementations don't support doing lookups on linear combinations of columns, but I figure they could support that if the savings are significant.

**Comment by mschof** [comment]
Yep, that's true. I remember that for Monolith in some cases we still included these linear columns for this reason, and in these cases the NTTs for those can be omitted.

I'm not sure if the approach can work with non-linear constraints -- wouldn't the multiplication need an NTT anyway?
